port/common/omrsysinfo_helpers.c

```diff
@@ -435,6 +435,35 @@ omrsysinfo_get_x86_description(struct OMRPortLibrary *portLibrary, OMRProcessorD
 	desc->features[3] = CPUInfo[CPUID_EBX]; /* Structured Extended Feature Flags in EBX */
 	desc->features[4] = CPUInfo[CPUID_ECX]; /* Structured Extended Feature Flags in ECX */
 
+	if (portLibrary->sysinfo_processor_has_feature(portLibrary, desc, OMR_FEATURE_X86_OSXSAVE)) {
+		/* Record extended processor state features from XCR0 */
+		unsigned int leaf = 0;
+#if defined(OMR_OS_WINDOWS)
+		unsigned long long xcr0 = _xgetbv(leaf);
+#else /* defined(OMR_OS_WINDOWS) */
+		unsigned int eax;
+      unsigned int edx;
+		__asm__ __volatile__("xgetbv" : "=a"(eax), "=d"(edx) : "c"(leaf));
+		unsigned long long xcr0 = ((unsigned long long) edx << 32) | eax;
+#endif /* defined(OMR_OS_WINDOWS) */
+
+		if (OMR_ARE_ANY_BITS_SET(xcr0, OMR_X86_XCR0_MASK_XMM)) {
+			portLibrary->sysinfo_processor_set_feature(portLibrary, desc, OMR_FEATURE_X86_XSAVE_SSE, TRUE);
+		}
+
+		if (OMR_ARE_ALL_BITS_SET(xcr0, OMR_X86_XCR0_MASK_XMM | OMR_X86_XCR0_MASK_YMM)) {
+			portLibrary->sysinfo_processor_set_feature(portLibrary, desc, OMR_FEATURE_X86_XSAVE_AVX, TRUE);
+		}
+
+		if (OMR_ARE_ALL_BITS_SET(xcr0, OMR_X86_XCR0_MASK_XMM | OMR_X86_XCR0_MASK_YMM | OMR_X86_XCR0_MASK_AVX512)) {
+			portLibrary->sysinfo_processor_set_feature(portLibrary, desc, OMR_FEATURE_X86_XSAVE_AVX512, TRUE);
+		}
+
+		if (OMR_ARE_ANY_BITS_SET(xcr0, OMR_X86_XCR0_MASK_APX_EGPR)) {
+			portLibrary->sysinfo_processor_set_feature(portLibrary, desc, OMR_FEATURE_X86_XSAVE_APX, TRUE);
+		}
+	}
```

Comment:
using `omrsysinfo_processor_set_feature` (primarily introduced for external use) is not optimal in this case. invoking it multiple times internally introduces unnecessary overhead. instead, consider constructing the full bitset using bitwise operations inside the new helper function, and directly set `desc->features[2]`. This approach is more efficient.

Reply:
I made this change, but I don't like how this tightly couples the code to the data structure.

Reply:
The purpose of `omrsysinfo_get_x86_description` is to directly populate the fields of `OMRProcessorDesc`. So, tightly coupling the code to the data structure is the goal here. If the preference is to use `omrsysinfo_processor_set_feature`, then it should be done externally (e.g. from `x/env/OMRCPU.cpp`).